### ISSUE TITLE
Inefficient and redundant implementation of aggregation states.

### ISSUE THREAD

**Issue by radek-kondziolka** [trinodb/trino#11820]
**The problem**

There are some redundant calls for aggregating operators that use more than one state.

Let's take the real example - `io.trino.operator.aggregation.RealAverageAggregation`

```
public class RealAverageAggregation
        extends SqlAggregationFunction
{
    ...
    public static void input(LongState count, DoubleState sum, long value)
    {
        count.setValue(count.getValue() + 1);
        sum.setValue(sum.getValue() + intBitsToFloat((int) value));
    }

    public static void combine(LongState count, DoubleState sum, LongState otherCount, DoubleState otherSum)
    {
        count.setValue(count.getValue() + otherCount.getValue());
        sum.setValue(sum.getValue() + otherSum.getValue());
    }

    public static void output(LongState count, DoubleState sum, BlockBuilder out)
    {
        if (count.getValue() == 0) {
            out.appendNull();
        }
        else {
            REAL.writeLong(out, floatToIntBits((float) (sum.getValue() / count.getValue())));
        }
    }
}

```
This aggregation function bases on two states - `LongState` and `DoubleState`. Let's call them atomic states. 
Currently, for an aggregation functions that uses more than one atomic state, `AccumulatorCompiler` wraps that set of atomic states to `RowType`. As a result, to make **groupped** accumulator state  `groupId` aware, the compiler generates `setGroupId` call (memory store) and `getGroupId` (memory load) for **every** atomic state.

It could be not so expensive but it is important to note that:

* `addInput` method is called per every row.  Let's how looks like generated code for:

```
   public void addInput(GroupByIdBlock groupIdsBlock, Page arguments, Optional mask) {
      this.state_0.ensureCapacity(groupIdsBlock.getGroupCount());
      this.state_1.ensureCapacity(groupIdsBlock.getGroupCount());
      Block masksBlock = (Block)mask.orElse((Object)null);
      Block block0 = arguments.getBlock(0);
      int rows = arguments.getPositionCount();
      int position = false;
      if (!AggregationUtils.maskGuaranteedToFilterAllRows(rows, masksBlock)) {
         for(int position = 0; CompilerOperations.lessThan(position, rows); ++position) {
            if (CompilerOperations.testMask(masksBlock, position) && !block0.isNull(position)) {
               this.state_0.setGroupId(groupIdsBlock.getGroupId(position));
               this.state_1.setGroupId(groupIdsBlock.getGroupId(position));
               this.state_0.input<invokedynamic>(this.state_0, this.state_1, block0, position);
            }
         }
      }

   }
```

* `addIntermediate` method method is called per every partial group:

```
      for(int position = 0; CompilerOperations.lessThan(position, rows); ++position) {
         if (!block.isNull(position) && !groupIdsBlock.isNull(position)) {
            this.state_0.setGroupId(groupIdsBlock.getGroupId(position));
            GroupedAccumulatorState var10000 = this.state_0;
            this.state_1.setGroupId(groupIdsBlock.getGroupId(position));
            GroupedAccumulatorState var10001 = this.state_1;
            this.stateSerializer_0.deserialize(columnBlock_0, position, (AccumulatorState)scratchState_0);
            this.stateSerializer_1.deserialize(columnBlock_1, position, (AccumulatorState)scratchState_1);
            var10000.combine<invokedynamic>(var10000, var10001, scratchState_0, scratchState_1);
         }
      }
```
* `evaluateIntermediate`, `evaluateFinal` are similar to the `addIntermediate` 

Especially, that problem rises when we are trying to rewrite decimal operators (sum, avg) to use atomic state. In that case we need to replace the dedicated state with three atomic states. As a result it is impossible to rewrite this without performance degredation. 

**Some results**

Currently, the decmial average function has a dedicated accumulator state (that means - not generated by a compiler): `LongDecimalWithOverflowAndLongState`. It was replaced by me with atomic states and collected microbenchmark `DecimalAverageAggregation.benchmark`. Actually, that metod benchmarks `addInput` method. This one is very performance sensitive because - let's say it one more time - it is executed per every row. 

The result is following:
```
(baseline) benchmarkAddInput   avg          1000       LONG  avgt    8      7.701           ns/op
(atomic)   benchmarkAddInput   avg          1000       LONG  avgt    8      9.960           ns/op
```

Let's take the "atomic" version of `avg` decimal function and apply change that can be described 
as: "Do not `setGroupId` and `getGroupId` for every atomic state. Instead of that, just pass 
`groupId` as an argument to the `DecimalAverageAggregation.inputLongDecimal` (and, respectively, 
add specialized methods for getting / setting state value, e.g.. By specialized metod I mean the additional methods that takes the `groupId` as an argument to get / save the value from / to an atomic state.
It is just only a demonstration version to show an negative impact of setting / getting `groupId` for every atomic state).
Let's look at result now:

```
(baseline)              benchmarkAddInput           avg              7.701           ns/op
(atomic)                benchmarkAddInput           avg              9.960           ns/op
(atomic+no_set_groupId) benchmarkAddInput           avg              7.287           ns/op
```

To sum up:
*There are some actually redundant memory stores and memory loads executed per row / partial groups. That redundancy is not free - there is a perofmance degradation. This is why that issue was created. In the next section there are presentend sketeches of some possible approaches.

**Propostion**

There could be some sketch of the approaches we can take here:

1. Add an annotation `@GroupId` and pass `groupId` as an argument to methods `addInput`, `addCombine` and so on.
That way makes operators `groupId`-aware. Now, operators are `groupId` agnostic. 
However, this change can be added without any change to existing operators. We could just rewrite some operators that are performance critical - like decimal operators. 
2. For every generated state (and only for them - unfortunately - we need to remember that we have much more dedicated states as well) we could inject special object by reference - something like `AccumulatorMetadata`. That object could track the `groupId` for all atomic states. 
Moreover, that `AccumulatorMetadata` object could keep the information about potential emptiness of the group - now even for empty (null) group `combine`, `serialize`, `desarialize` methods are called. In a result the `combine` method is complicated - bigger, slower, more hardly inlineable - for decimal sum it was proved that it could much faster (like 1.5x - 2x times).

This issue should just open the discussion and draw our attention to it.

**Comment by sopel39** [comment]
cc @martint @dain

**Comment by martint** [comment]
Option #1 is reasonable. Option #2 is too complicated.

Also, we're going to need to make the group id explicit in the calls anyway to be able to support enhanced aggregation implementations that can use vectorization APIs, GPU, etc.

**Comment by radek-starburst** [comment]
@martint ,

> Also, we're going to need to make the group id explicit in the calls anyway to be able to support enhanced aggregation > implementations that can use vectorization APIs, GPU, etc.

do you think that we should have:

1. a dedicated operators for ungroupped / groupped operators?:
    
```
    public class GroupedDecimalAverageAggregation {
         public static void input(LongState state, @GroupId groupId);
    }

    
    public class SingleDecimalAverageAggregation {
         public static void input(LongState state);
    }
```

2. a one operator with dedicated methods?:
    
```
    public class DecimalAverageAggregation {
         public static void input(LongState state, @GroupId groupId);
    }

    
    public class DecimalAverageAggregation {
         public static void input(LongState state);
         public static void input(LongState state, @GroupId groupId);
    }
```

and make it possible to that change would be backward-compatibe to avoid making a revolution at once (it is easily possible).

**Comment by dain** [comment]
Option 1 is not reasonable.  This is not how aggregations work and would rewrite redesigning the framework.  
Option 2 I don't understand but sounds complicated.

If your goal is to reduce the extra calls in generated aggregation code, then then you could generate a single aggregation state the encodes all nested states into one object.  This could help slightly but honestly, I don't think this minor performance change is worth any effort.

**Comment by dain** [comment]
BTW, my guess is the performance change is more related to explicitly passing the groupId value than it is to merging the two states together.  Under the covers these grouped states are just arrays, and passing the array index value via a parameter is likely to be faster then dereferencing the array index from a field in the object.... BTW, just a guess

**Comment by martint** [comment]
> and passing the array index value via a parameter is likely to be faster then dereferencing the array index from a field in the object...

That's exactly what option 1 is about -- pass the group id explicitly as an argument to the method instead of calling `setGroupId` first and then calling the method.

**Comment by radek-starburst** [comment]
1. Yes - but, please note that we are going to rewrite dedicated state `LongDecimalWithOverflowAndLongState` to 3 atomic states. When we do it, we have two extra call for every row, partial group and so on. 
2. The second issue: When we want to have an explicit groupId argument we need to have two methods:
* `addInput(state, .., groupId)` (and same for other aggregating methods) for *grouping* case
* `addInput(state, ..)` (and same for other aggregating methods) for *not* grouping case
Moreover, we need to have a additional interface, like:
```
interface LongState {
   void setValue(long value);
} 
```
```interface GroupedLongState {
void setValue(long groupId, long value);
} 
```

but only for some operators.

It is doeable by kind of extension and changes in compilers. I am aware that it's kind to the framework and this is why I've created that issue. 
Nevertheless. if we want to rewrite `LongDecimalWithOverflowAndLongState` to atomic states without performance degredation we have to avoid extra `setGroupId` per every row.

**Comment by radek-starburst** [comment]
> If your goal is to reduce the extra calls in generated aggregation code, then then you could generate a single  aggregation state the encodes all nested states into one object. This could help slightly but honestly, I don't think this minor performance change is worth any effort.

If I understand you correctyly, this what `LongDecimalWithOverflowAndLongState` do (it written not generated) but it implements that idea.

**Comment by sopel39** [comment]
> The second issue: When we have an explicit groupId argument we need to have two methods:

I'm not sure you need two methods and two interfaces. It could be that for non partitioned case, group id is just 0.

**Comment by radek-starburst** [comment]
> I'm not sure you need two methods and two interfaces. It could be that for non partitioned case, group id is just 0.
Yes, I was thinking about it. Then, you have the inteface like:

```
interface LongState {
    void setValue(long groupId, long value);
    long getValue(long groupId);
}

class SingleLongState implements LongState {
    void setValue(long groupId, long value) {
           this.value = value; // ignore groupId
     }
    long getValue(long groupId) {
         return this.value;
    }

class SingleLongState implements LongState {
    void setValue(long groupId, long value) {
           // set the element in the array by groupId
     }
    long getValue(long groupId) {
           // get the element from the array by groupId
    }
}

```
I decided that it is too confusing for our users of API (in the future). For internal purpose it is acceptable for me (and good approach). But, it is a debatable issue, that approach is simpler and more elegant. What do you think @dain , @martint ?

**Comment by dain** [comment]
I have been thinking about this for a while now, and think there is an approach that is backward and not too complex (maybe you all have already thought of this).

I believe the proposal is to convert the function signature from:
```
void input(LongState count, DoubleState sum, long value);
void combine(LongState count, DoubleState sum, LongState otherCount, DoubleState otherSum);
void output(LongState count, DoubleState sum, BlockBuilder out);
```
to
```
void input(@GroupId long groupId, LongState count, DoubleState sum, long value);
void combine(@GroupId long groupId, LongState count, DoubleState sum, @GroupId long groupId, LongState otherCount, DoubleState otherSum);
void output(@GroupId long groupId, LongState count, DoubleState sum, BlockBuilder out);
```
We could adapt the old interface to the new interface using basic method handle combinators.   Basically we use combinators to process the args as follows:
```
void input(@GroupId long groupId, LongState count, DoubleState sum, long value)
{
    count.setGroupId(groupId);
    sum.getGroupId(groupId);
    oldStyleInput(count, sum, value);
}
```
This can be done completely without byte code generation.  

So, assuming we want to make this change, we would change the AggregationMetadata to expect the new method signature, and obviously change the aggregation compiler to expect invoke the new signature.  Finally, the AggregationFromAnnotationsParser would be changed to detect the old and new patterns.  When the old is detected the method handle would be adapted to the new signature (using combinators).  Finally, aggregations would be switched to the new style.  This work should be based on the PR i have up that rewrites all (except for reduce) aggregations to annotated forms, which will make this easier.  Also, that PR adds a number of new features that would conflict with this change.

**Comment by radek-starburst** [comment]
Super, it solves the problem of `input/combine/output` interfaces. 

Now, we have to imagine how our "atomic" states interfaces should look like. We have two methods:

`oldstyle` one:
```
    void oldStyleInput(LongState count, LongState sum, long value) {
         count.setValue(count.getValue() + 1); // We have the method without `groupId`
    }

``` 
`newstyle` one:
```
    void oldStyleInput(LongState count, LongState sum, long value, @GroupId int groupId) {
         count.setValue(groupId, count.getValue(groupId) + 1); // We have the method with `groupId` given
    }
```

Potentially, we would have the interface:

```
class LongState {
    void setValue(@GroupId int groupId, long value); 
    void setValue(long value); 
    ...
 }
```

and generated implementation of `SingleLongState` could just throw the exception `UnsupportedOperation` for grouped operations. But we have another problem: the `combine` takes two parameters: **grouped** state for the leftside, and **single** state for the rightsite - I mean:


```
    void combine(@GroupId int groupId, LongState lhs, LongState rhs) {
          lhs.setValue(groupId, lhs.getGroupId(groupId) + rhs.getValue(groupId)); 
    }
```
This method causes the `UnsupportedException` because `rhs` is actually `Single` state. Especially, it would be very misleading and confusing for our potential users of our API. 

These are my rationale and this is why I decided that explicit states for group-id aware states should be separated states.
It is not nice, super seamless and so on. It is possible that I cannot find here any better idea to implement it.

**Comment by sopel39** [comment]
> and generated implementation of SingleLongState could just throw the exception UnsupportedOperation for grouped operations.

Why would it fail? Maybe it could just assert group id is 0

**Comment by radek-starburst** [comment]
> Why would it fail? Maybe it could just assert group id is 0

It is one option. However, `combine` is still problematic. Anyway, if do **not** assert `groupId == 0` for `SingleState` that approach works and it resolves all problems I see (for now).

It looks like we have a kind of common version. To sum up our discussion, our proposal changes could be expressed as:

**Make aggregation operators in Trino group oriented instead of not-group oriented. **

Then, basically we have a such state:
```
interface LongState {
    void setValue(int groupId, long value);
```
and similarly for serializer / deserializers. Single states implementations has `groupId` param, but they ignore that.
It looks like it is what we agreed here, do we? @dain @martint @sopel39 ?

**Comment by dain** [comment]
Yes.  Also since we are effectively introducing a v2 of the aggregation annotations, we should do a review of the APIs here since anyone using v2 has to rewrite their code.  I say this because processing the very flexible super-legacy annotated aggregations makes the code very complex, so the more we can do to say "it must be exactly like this" makes this stuff easier to maintain.

Specifically, we should require full annotation of the parameters aggregation, as I did rewrite PR I mentioned above.  In that PR, I added support for having multiple accumulator states, and for the aggregations that switched to that mode, I added tighter controls over annotation requirements, and argument ordering.  For example, I would require the argument for input method to be:
1. Injected dependencies (operators, types, etc)
2. `@GroupId` 
3. `@AggregationState`
4. `@SqlType` input parameters
I would have similar rules for combine and output.  Similarly, for states and serializers.  If we find the rules too burdensome, we can relax them later, but this is our one chance to tighten things up.

**Comment by radek-starburst** [comment]
Sketch of the APIv2: 
### AggregationFunction

We annotate aggregation function by @AggregationFunction. Possibly, we could add param `explicitGroupId` (by default `false`) to that annotation to express that this operators is going to be groupId-oriented. On my eye it would facilitate to write generators (`StateCompiler`, `AccumulatorCompiler`) and, especially, to validate functions. Otherwisate, we have to recognize the @GroupId param in states, in method params, and son on.   

- input 

```
@InputFunction
void input(`Injected dependencies (operators, types, etc)`, @GroupId long groupId, @AggregationState AccumulatorState state ..., @SqlType,  @SqlType input parameters, @BlockPosition @SqlType("TPYE") Block value, @BlockIndex int position);
```
some _params_ in the `input` are not required, like dependencies, `value`, `position` and so on.


- combine 

```
@CombineFunction
void combine(@GroupId long groupId, @AggregationState AccumulatorState state);
```

- output

```
@OutputFunction("TYPE")
void output(@TypeParameter("TYPE") Type type, @GroupId long groupId, @AggregationState AccumulatorState state, BlockBuilder out);
```
__type__ is not required

### Serializer / Deserializer

```
public interface AccumulatorStateSerializer<T extends AccumulatorState>
{
    Type getSerializedType();

    void serialize(@GroupId long groupId, T state, BlockBuilder out);
    void deserialize(@GroupId long groupId, Block block, int index, T state);
}
```

### AccumulatorState

```
public interface AccumulatorState
{
    long getEstimatedSize();

    default AccumulatorState copy()
    {
        throw new UnsupportedOperationException("copy not implemented for " + getClass());
    }
}
```
class LongState implements AccumulatorState {
      long getValue(@GroupId long groupId);
      void setValue(@GroupId long groupId, long value);
}
```

**Comment by sopel39** [comment]
> Specifically, we should require full annotation of the parameters aggregation, as I did rewrite PR I mentioned above. In that PR, I added support for having multiple accumulator states, and for the aggregations that switched to that mode, I added tighter controls over annotation requirements, and argument ordering. For example, I would require the argument for input method to be:

@dain should we wait for that PR to land first? Is this https://github.com/trinodb/trino/pull/12588?

**Comment by dain** [comment]
@radek-starburst some notes:
* All functions support `Injected dependencies (operators, types, etc)`, most don't use them.  For example, in the output function you show `@TypeParameter("TYPE") Type type` which is a type dependency.
* Block and position shown in the first input is just another argument calling convention, so same thing as parameters.
* You forgot about the remove function, but it has the same shape as input.
* AccumulatorStateSerializer doesn't need the `@GroupId` annotation on the argument since this is an interface we provide that the function provider implements
* For the AccumulatorState, I don't think I would require the `@GroupId` annotation either, but I'm not sure here.  Unlike the main aggregation functions, it is very clear which argument is the group id, but it might be nice for readability... I don't have a strong preference either way.
* The new InOut state (see `MaxAggregationFunction`) will need to be modified as part of this. This will require changes to the `IN_OUT` calling convention, and the code `ScalarFunctionAdapter`.  This is a very complex change.  I think we may be able to run these functions in "legacy" mode until, someone that really knows Java method handle combinators can update that code.  I expect that this will be me, but I'm happy to just be a reviewer :)

@sopel39 Yes I meant #11477.  My new PR #12588 moves the function interface to the SPI.  Before we make the new interfaces non-experimental, we need to finish this work because there is no way to make this a backwards compatible change for the low level APIs.... so to be super clear, if we decide to land #12588 before this work is done, we must make clear that the function APIs are experimental and will change.  Once this work is finalized and landed, then we could make function APIs final (non-experimental).

**Comment by findepi** [comment]
do @dain 's recent changes to aggregate states make this issue obsolete?

**Comment by radek-starburst** [comment]
I do not think so - I do not see how they improve partial representation of aggregation states.
